python/pyarrow/tests/test_fs.py

```diff
@@ -0,0 +1,266 @@
+        id='LocalFileSystem'
+    ),
+    pytest.param(
+        lambda tmp: SubTreeFileSystem(tmp, LocalFileSystem()),
+        id='SubTreeFileSystem(LocalFileSystem)'
+    ),
+    # s3 follows
+    # pytest.param(
+    #     lambda tmp: SubTreeFileSystem(tmp, LocalFileSystem()),
+    #     name='SubTreeFileSystem[LocalFileSystem]'
+    # ),
+])
+def fs(request, tempdir):
+    return request.param(tempdir)
+
+
+@pytest.fixture(params=[
+    pytest.param(pathlib.Path, id='Path'),
+    pytest.param(str, id='str')
+])
+def testpath(request, fs, tempdir):
+    # we always use the tempdir for rewding and writing test artifacts, but
+    # if the filesystem is wrapped in a SubTreeFileSystem then we don't need
+    # to prepend the path with the tempdir, we also test the API with both
+    # pathlib.Path objects and plain python strings
+    if isinstance(fs, SubTreeFileSystem):
+        return lambda path: request.param(path)
+    else:
+        return lambda path: request.param(tempdir / path)
+
+
+def test_stat_with_paths(fs, tempdir, testpath):
+    (tempdir / 'a' / 'aa' / 'aaa').mkdir(parents=True)
+    (tempdir / 'a' / 'bb').touch()
+    (tempdir / 'c.txt').touch()
+
+    aaa_path = testpath('a/aa/aaa')
+    bb_path = testpath('a/bb')
+    c_path = testpath('c.txt')
+
+    aaa, bb, c = fs.stat([aaa_path, bb_path, c_path])
+
+    assert aaa.path == pathlib.Path(aaa_path)
+    assert aaa.base_name == 'aaa'
+    assert aaa.extension == ''
+    assert aaa.type == FileType.Directory
+
+    assert bb.path == pathlib.Path(bb_path)
+    assert bb.base_name == 'bb'
+    assert bb.extension == ''
+    assert bb.type == FileType.File
+
+    assert c.path == pathlib.Path(c_path)
+    assert c.base_name == 'c.txt'
+    assert c.extension == 'txt'
+    assert c.type == FileType.File
+
+    selector = Selector(testpath(''), allow_non_existent=False, recursive=True)
+    nodes = fs.stat(selector)
+    assert len(nodes) == 5
+    assert len(list(n for n in nodes if n.type == FileType.File)) == 2
+    assert len(list(n for n in nodes if n.type == FileType.Directory)) == 3
+
+
+def test_mkdir(fs, tempdir, testpath):
+    directory = testpath('directory')
+    directory_ = tempdir / 'directory'
+    assert not directory_.exists()
+    fs.mkdir(directory)
+    assert directory_.exists()
+
+    # recursive
+    directory = testpath('deeply/nested/directory')
+    directory_ = tempdir / 'deeply' / 'nested' / 'directory'
+    assert not directory_.exists()
+    with pytest.raises(ArrowIOError):
+        fs.mkdir(directory, recursive=False)
+    fs.mkdir(directory)
+    assert directory_.exists()
+
+
+def test_rmdir(fs, tempdir, testpath):
+    folder = testpath('directory')
+    nested = testpath('nested/directory')
+    folder_ = tempdir / 'directory'
+    nested_ = tempdir / 'nested' / 'directory'
+
+    folder_.mkdir()
+    nested_.mkdir(parents=True)
+
+    assert folder_.exists()
+    fs.rmdir(folder)
+    assert not folder_.exists()
```

Comment:
That won't work with non-local filesystems.

Reply:
I'll write other test cases than.